### DISCUSSION TITLE
Give integrations access to routes earlier

### DISCUSSION THREAD

**Discussion by florian-lefebvre** [withastro/roadmap#998] [Proposal]
### Body

# Summary

Give integrations access to routes earlier

# Background & Motivation

Currently, routes are only exposed in `astro:build:done`. However, if you want as an integration author to do some stuff relying on routes earlier in the lifecycle like in dev, or for typegen, you basically can't.

The proposal is to expose the data earlier.

Something to keep in mind is that user routes (not injected) can change during dev so some kind of watcher may need to be exposed.

# Goals

- Have access to routes earlier in the lifecycle
- Nice to have: a way to listen to user routes changes in dev

# Example

When loading an astro project, no route is added after `astro:config:setup` so the initial route data could be exposed on `astro:config:done`. As for the watcher, it could be on `astro:server:setup` alongside the vite dev server.

**Comment by florian-lefebvre** [comment]
@Fryuni feel free to give your thoughts on this! FYI https://github.com/withastro/astro/pull/11824 should allow exposing route data on `astro:config:done`

**Comment by Fryuni** [comment]
Whichever hook this information is made available would need to be re-run during the execution of the dev server or would need to receive a function for a callback like so:
```ts
{
  'astro:config:done': ({ routes, onRouteChanged }) => {
    // Do something with the routes
    onRouteChanged(newRoutes => {
      // Do something with the changed routes due to an edit while running the dev server
    });
  },
}
```

Instead of that, I think this would be better as a separate hook:
```ts
  `astro:routes:resolved`: ({ routes }) => {
    // This hook is called whenever the routes list is resolved.
    // If it changes, it will be called again
  }
```

It follows the same idea as the recently added `astro:route:setup` but for the entire routes list.

As for the information in this hook, I'd like it to have at least:
- The route (e.g. `/[...slug]` or `/blog/[id]`). Something that is currently missing from the `astro:route:setup`.
- The entrypoint source code, which is currently called `component` in the inner type but is not necessarily a component.
- Whether it is pre-rendered or not.

A nice thing to have would be an `origin`/`source` field indicating whether it came from the user project or was injected from an integration.

**Reply by targetlucked69** [reply]
Will you have access to the astro/vite config in the proposed `astro:routes:resolved` hook so that you can for example create a virtual module that exports a map of routes, register the plugin to Vite, and invalidate them as well if route changes?

**Reply by ematipico** [reply]
@walmartwarlord the config is already exposed by other hooks, so no need to add the same data again

**Comment by florian-lefebvre** [comment]
If there's a new hook, it will run after `astro:config:setup` so you won't be able to use this data inside a vite plugin...unless maybe you use [inline mode](https://inox-tools.fryuni.dev/inline-mod/lazy) from Luiz

**Reply by Fryuni** [reply]
> so you won't be able to use this data inside a vite plugin

Ye of little faith!

It runs after the plugin is added to Vite, but before Vite runs (which is after `astro:build:start`/`astro:server:start`).

Just do this:
```ts
const routeData: RouteData[] = []; // This runs first

return {
  hooks: {
    'astro:config:setup': (params) => {
      addVitePlugin(params, { // This runs second
        plugin: {
          load(id) { // This runs fourth
            // Use the `routeData` from above
          }
        }
      });
    },
    'astro:routes:resolved': ({routes}) => { // This runs third
      // Clear previous
      routeData.splice(0, routeData.length);
      // Add the new data
      routeData.push(...routes);
    },
  }
};
```

**Reply by Fryuni** [reply]
Or you can use Inline Mod, sure. But it is overkill for this 😆

**Reply by florian-lefebvre** [reply]
oh i didn't know!

**Comment by matthewp** [comment]
I don't mind this in theory. Have you looked at what code changes would need to be made for this? In particular, for dev mode I guess we'd need access to run the hook on route changes. Do we have a place for that?

**Reply by Princesseuh** [reply]
In `next`, I don't think any code change is required since we do the manifest up front now

**Reply by florian-lefebvre** [reply]
That would calling here https://github.com/withastro/astro/blob/main/packages/astro/src/core/routing/manifest/create.ts#L727, i'm looking at how it works in dev now.

EDIT: this function is called in dev as well so that's really the only place where we need to add it

**Comment by matthewp** [comment]
@florian-lefebvre if we add this route can we also deprecate providing the routes to `astro:build:done`?

**Reply by florian-lefebvre** [reply]
I guess yes

**Comment by florian-lefebvre** [comment]
PR at https://github.com/withastro/astro/pull/12329

**Comment by florian-lefebvre** [comment]
This should be merged soon, I'm closing as accepted
